File: desktop/src/surface/adapters/plans.ts
import type { EvidenceLink, FigureGrade, GoalAccountView, GoalHistoryView, GoalPlanView, GoalProposalView, PlanDraftResult, PlanDraftView, PlanPayload, PlanVerb, PlansData } from "../types";
import { isRecord, textValue } from "./primitives";
import { adaptActionOutcome } from "./questions";

const VERBS: readonly PlanVerb[] = ["create", "change_terms", "reserve", "release", "pause", "resume", "set_aside"];
const GRADES: readonly FigureGrade[] = ["verified", "corroborated", "unverified", "conflicted", "unavailable", "not_applicable"];
const STATUSES: readonly GoalPlanView["status"][] = ["complete", "paused", "ahead", "on_track", "at_risk", "unscheduled"];
const GROUPS: readonly GoalPlanView["group"][] = ["active", "paused", "complete", "set_aside"];
const STATES: readonly GoalPlanView["state"][] = ["active", "paused", "set_aside"];

function strings(raw: unknown): readonly string[] { return Array.isArray(raw) ? raw.map(textValue).filter(Boolean) : []; }
function payload(raw: unknown): PlanPayload {
  if (!isRecord(raw)) return {};
  return Object.fromEntries(Object.entries(raw).filter(([, value]) => typeof value === "string" || typeof value === "number" || value === null)) as PlanPayload;
}
function links(raw: unknown): readonly EvidenceLink[] {
  if (!Array.isArray(raw)) return [];
  return raw.filter(isRecord).map((row) => ({ targetDocumentId: textValue(row.document_id), label: textValue(row.label), relation: textValue(row.relation) as EvidenceLink["relation"], page: textValue(row.page) })).filter((row) => row.targetDocumentId);
}
function account(raw: unknown): GoalAccountView | null {
  if (!isRecord(raw) || !textValue(raw.id) || typeof raw.eligible !== "boolean") return null;
  const grade = GRADES.find((word) => word === raw.grade) ?? "";
  return { id: textValue(raw.id), name: textValue(raw.name), currency: textValue(raw.currency), eligible: raw.eligible, balance: textValue(raw.balance), balanceDisplay: textValue(raw.balance_display), reserved: textValue(raw.reserved), reservedDisplay: textValue(raw.reserved_display), available: textValue(raw.available), availableDisplay: textValue(raw.available_display), grade, gradeDescription: textValue(raw.grade_description), dated: textValue(raw.dated), asOf: textValue(raw.as_of), balanceExplanation: textValue(raw.balance_explanation), sourceDocumentId: textValue(raw.source_document_id), sourcePage: textValue(raw.source_page), sourceRegion: textValue(raw.source_region), sourceNote: textValue(raw.source_note), caveats: strings(raw.caveats), sentence: textValue(raw.sentence), reason: textValue(raw.reason), evidenceLinks: links(raw.citations) };
}
function history(raw: unknown): GoalHistoryView | null {
  if (!isRecord(raw) || (raw.kind !== "reserved" && raw.kind !== "released")) return null;
  return { kind: raw.kind, accountId: textValue(raw.account_id), amount: textValue(raw.amount), amountDisplay: textValue(raw.amount_display), reason: textValue(raw.reason), occurredAt: textValue(raw.occurred_at), sentence: textValue(raw.sentence), valid: raw.valid === true };
}
function goal(raw: unknown): GoalPlanView | null {
  if (!isRecord(raw)) return null;
  const group = GROUPS.find((word) => word === raw.group);
  const state = STATES.find((word) => word === raw.state);
  const status = STATUSES.find((word) => word === raw.status);
  if (!textValue(raw.id) || !textValue(raw.title) || !group || !state || !status) return null;
  const accounts = (Array.isArray(raw.accounts) ? raw.accounts : []).map(account).filter((row): row is GoalAccountView => row !== null);
  const rows = (Array.isArray(raw.history) ? raw.history : []).map(history).filter((row): row is GoalHistoryView => row !== null);
  const actions = (Array.isArray(raw.actions) ? raw.actions : []).map((word) => VERBS.find((candidate) => candidate === word)).filter((word): word is PlanVerb => Boolean(word));
  return { id: textValue(raw.id), title: textValue(raw.title), group, state, status, statusLabel: textValue(raw.status_label), headline: textValue(raw.headline), explanation: textValue(raw.explanation), currency: textValue(raw.currency), targetAmount: textValue(raw.target_amount), targetDisplay: textValue(raw.target_display), targetDate: textValue(raw.target_date), reserved: textValue(raw.reserved), reservedDisplay: textValue(raw.reserved_display), remaining: textValue(raw.remaining), remainingDisplay: textValue(raw.remaining_display), monthlyContribution: textValue(raw.monthly_contribution), monthlyDisplay: textValue(raw.monthly_display), contributionDay: typeof raw.contribution_day === "number" ? raw.contribution_day : null, requiredMonthly: textValue(raw.required_monthly), requiredMonthlyDisplay: textValue(raw.required_monthly_display), projectedCompletionDate: textValue(raw.projected_completion_date), deviation: textValue(raw.deviation), deviationDisplay: textValue(raw.deviation_display), nextContributionDate: textValue(raw.next_contribution_date), noMoneyMoved: textValue(raw.no_money_moved), accounts, history: rows, historyNote: textValue(raw.history_note), assumptions: strings(raw.assumptions), caveats: strings(raw.caveats), actions };
}
function proposal(raw: unknown): GoalProposalView | null {
  if (!isRecord(raw)) return null;
  const verb = VERBS.find((word) => word === raw.verb);
  if (!textValue(raw.id) || !verb) return null;
  const actions = Array.isArray(raw.actions) ? raw.actions.filter((word): word is "confirm" | "decline" => word === "confirm" || word === "decline") : [];
  const display = isRecord(raw.display) ? Object.fromEntries(Object.entries(raw.display).map(([key, value]) => [key, textValue(value)]).filter(([, value]) => value)) : {};
  return { id: textValue(raw.id), verb, goalId: textValue(raw.goal_id), summary: textValue(raw.summary), consequence: textValue(raw.consequence), noMoneyMoved: textValue(raw.no_money_moved), exact: payload(raw.exact), display, assumptions: strings(raw.assumptions), actions };
}

export function adaptPlans(raw: unknown): PlansData | null {
  if (!isRecord(raw) || !isRecord(raw.invitation) || !["absent", "ready", "partial"].includes(textValue(raw.state))) return null;
  const goals = (Array.isArray(raw.goals) ? raw.goals : []).map(goal).filter((row): row is GoalPlanView => row !== null);
  const proposals = (Array.isArray(raw.proposals) ? raw.proposals : []).map(proposal).filter((row): row is GoalProposalView => row !== null);
  const groups = isRecord(raw.groups) ? Object.fromEntries(Object.entries(raw.groups).map(([key, value]) => [key, strings(value)])) : {};
  return { state: raw.state as PlansData["state"], title: textValue(raw.title), invitation: { title: textValue(raw.invitation.title), body: textValue(raw.invitation.body) }, noMoneyMoved: textValue(raw.no_money_moved), goals, groups, proposals, actions: ["draft", "propose", "confirm", "decline"] };
}

export function adaptPlanDraftView(raw: unknown): PlanDraftView | null {
  if (!isRecord(raw)) return null;
  const verb = VERBS.find((word) => word === raw.verb);
  return verb && isRecord(raw.payload) && isRecord(raw.calculated) ? { verb, payload: payload(raw.payload), calculated: Object.fromEntries(Object.entries(raw.calculated).map(([key, value]) => [key, textValue(value)])) } : null;
}

export function adaptPlanDraftReply(raw: unknown): PlanDraftResult {
  if (!isRecord(raw) || !["completed", "waiting", "refused"].includes(textValue(raw.kind)) || !textValue(raw.message) || !isRecord(raw.state)) return { state: "unreadable" };
  const outcome = adaptActionOutcome(raw);
  if (!outcome) return { state: "unreadable" };
  const row = raw.state;
  const draftState = textValue(row.draft_state);
  if (!["ready", "needs_input", "refused"].includes(draftState)) return { state: "unreadable" };
  const draft = adaptPlanDraftView(row.draft);
  const kind = draftState === "ready" ? "ready" : draftState === "needs_input" ? "needs_input" : "refused";
  return { state: "settled", kind, message: outcome.message, reason: outcome.reason, draft };
}

export function adaptPlanActionOutcome(raw: unknown) {
  return adaptActionOutcome(raw);
}

File: product/viva/surface/plans.py
"""Reviewed GoalsAndPlans.v1 read model; the desktop only renders it."""

from __future__ import annotations

from decimal import Decimal

from .. import render
from ..persona import STOOD_BEHIND_MOMENT, moment
from .models import (Citation, CitationRelation, GoalAccountView,
                     GoalHistoryView, GoalPlanView, GoalProposalView)


_EXCLUSIONS = {
    "account_not_depository": "plans_exclusion_not_depository",
    "account_not_issuer": "plans_exclusion_not_issuer",
    "account_currency_unstated": "plans_exclusion_currency",
    "account_currency_differs": "plans_exclusion_currency",
    "account_balance_conflicted": "plans_exclusion_conflicted",
}


def _account(projection, row, locale: str) -> GoalAccountView:
    record_id = row.provenance.doc_id
    documents = projection.captured_docs()
    evidence = (record_id,) if record_id and record_id in documents else ()
    citations = tuple(Citation(
        document_id=item,
        label=projection.captured_filenames().get(item, ""),
        page=str(row.provenance.page or ""),
        relation=CitationRelation.ATTESTS) for item in evidence)
    return GoalAccountView(
        id=row.account_id, name=row.name, currency=row.currency,
        eligible=True, balance=str(row.balance),
        balance_display=render.money(row.balance, row.currency, locale=locale),
        reserved=str(row.reserved),
        reserved_display=render.money(row.reserved, row.currency, locale=locale),
        available=str(row.available),
        available_display=render.money(row.available, row.currency, locale=locale),
        grade=row.grade,
        grade_description=moment(STOOD_BEHIND_MOMENT + row.grade),
        dated=row.dated, as_of=str(row.as_of or ""),
        balance_explanation=row.explanation,
        source_document_id=record_id,
        source_page=str(row.provenance.page or ""),
        source_region=row.provenance.region,
        source_note=row.provenance.note,
        caveats=(row.explanation,) if row.explanation else (),
        sentence=moment(
            "plans_account_available",
            balance=render.money(row.balance, row.currency, locale=locale),
            reserved=render.money(row.reserved, row.currency, locale=locale),
            available=render.money(row.available, row.currency, locale=locale)),
        evidence_ids=evidence, citations=citations)


def _excluded(row) -> GoalAccountView:
    return GoalAccountView(
        id=row.account_id, name=row.name, currency=row.currency,
        eligible=False, reason=row.reason,
        sentence=moment(_EXCLUSIONS[row.reason]))


def _history(row, currency: str, locale: str) -> GoalHistoryView:
    amount = row.applied_amount
    return GoalHistoryView(
        kind=row.kind, account_id=row.account_id, amount=str(amount),
        amount_display=render.money(amount, currency, locale=locale),
        reason=row.reason, occurred_at=row.occurred_at,
        sentence=moment(
            "plans_history_reserved" if row.kind == "reserved"
            else "plans_history_released",
            amount=render.money(amount, currency, locale=locale)),
        valid=row.valid)


def _actions(row) -> tuple[str, ...]:
    if row.state == "set_aside":
        return ()
    actions = ["change_terms"]
    if row.state == "active":
        actions.append("pause")
    elif row.state == "paused":
        actions.append("resume")
    if row.available_accounts and row.state != "set_aside":
        actions.append("reserve")
    if row.reserved:
        actions.append("release")
    elif row.state != "set_aside":
        actions.append("set_aside")
    return tuple(actions)


def _goal(projection, row, locale: str) -> GoalPlanView:
    monthly = row.monthly_contribution
    required = row.required_monthly
    deviation = row.deviation
    group = "complete" if row.status == "complete" else row.state
    accounts = tuple(_account(projection, item, locale)
                     for item in row.available_accounts)
    accounts += tuple(_excluded(item) for item in row.exclusions)
    return GoalPlanView(
        id=row.goal_id, title=row.title, group=group, state=row.state,
        status=row.status, status_label=moment("plans_status_" + row.status),
        headline=moment("plans_goal_headline"),
        explanation=moment("plans_goal_explanation"), currency=row.currency,
        target_amount=str(row.target_amount),
        target_display=render.money(
            row.target_amount, row.currency, locale=locale),
        target_date=row.target_date, reserved=str(row.reserved),
        reserved_display=render.money(row.reserved, row.currency, locale=locale),
        remaining=str(row.remaining),
        remaining_display=render.money(
            row.remaining, row.currency, locale=locale),
        monthly_contribution=str(monthly) if monthly is not None else "",
        monthly_display=(render.money(monthly, row.currency, locale=locale)
                         if monthly is not None else ""),
        contribution_day=row.contribution_day,
        required_monthly=str(required) if required is not None else "",
        required_monthly_display=(
            render.money(required, row.currency, locale=locale)
            if required is not None else ""),
        projected_completion_date=row.projected_completion_date,
        deviation=str(deviation) if deviation is not None else "",
        deviation_display=(render.money(
            deviation, row.currency, locale=locale)
            if deviation is not None else ""),
        next_contribution_date=row.next_contribution_date,
        no_money_moved=moment("plans_local_reservation"),
        accounts=accounts,
        history=tuple(_history(item, row.currency, locale)
                      for item in row.history),
        history_note=(moment("plans_history_withheld") if row.issues else ""),
        assumptions=(moment("plans_goal_explanation"),),
        caveats=((moment("plans_history_withheld"),) if row.issues else ()),
        event_ids=row.event_ids,
        actions=_actions(row))


def _proposal(projection, row: dict, locale: str, today: str) -> GoalProposalView:
    exact = dict(row.get("proposal") or {})
    verb = str(row.get("verb") or "")
    currency = str(exact.get("currency") or "")
    goal = None
    if not currency and exact.get("goal_id"):
        try:
            goal = projection.goal(str(exact["goal_id"]), today)
            currency = goal.currency
        except KeyError:
            currency = ""
    amount_text = exact.get("target_amount") \
        if verb in ("create", "change_terms") else exact.get("amount")
    fields = {}
    if amount_text not in (None, ""):
        fields["amount"] = render.money(
            Decimal(str(amount_text)), currency, locale=locale)
    display = {}
    title = str(exact.get("title") or getattr(goal, "title", ""))
    if title:
        display["plan_name"] = title
    account_id = str(exact.get("account_id") or "")
    if goal is not None and account_id:
        accounts = (*goal.available_accounts, *goal.exclusions)
        account = next((item for item in accounts
                        if item.account_id == account_id), None)
        if account is not None:
            display["account_name"] = account.name
    if exact.get("target_amount") not in (None, ""):
        display["target_amount"] = render.money(
            Decimal(str(exact["target_amount"])), currency, locale=locale)
    if exact.get("amount") not in (None, ""):
        display["amount"] = render.money(
            Decimal(str(exact["amount"])), currency, locale=locale)
    if exact.get("monthly_contribution") not in (None, ""):
        display["monthly_contribution"] = render.money(
            Decimal(str(exact["monthly_contribution"])), currency,
            locale=locale)
    for name in ("target_date", "contribution_day"):
        if exact.get(name) not in (None, ""):
            display[name] = str(exact[name])
    return GoalProposalView(
        id=str(row.get("proposal_id") or ""), verb=verb,
        goal_id=str(exact.get("goal_id") or ""),
        summary=moment("plans_proposal_" + verb, **fields),
        consequence=moment("plans_proposal_consequence"),
        no_money_moved=moment("plans_local_reservation"), exact=exact,
        display=display,
        assumptions=(moment("plans_goal_explanation"),))


def plans(projection, locale: str, today: str) -> dict:
    """Compose the durable plan workspace from one projection read."""
    goals = tuple(_goal(projection, row, locale)
                  for row in projection.goals(today))
    proposals = tuple(_proposal(projection, row, locale, today)
                      for row in projection.open_goal_proposals())
    if not goals and not proposals:
        state = "absent"
    elif any(row.caveats for row in goals):
        state = "partial"
    else:
        state = "ready"
    return {
        "state": state, "title": moment("plans_title"),
        "invitation": {
            "title": moment("plans_empty_title"),
            "body": moment("plans_empty_body"),
        },
        "no_money_moved": moment("plans_local_reservation"),
        "goals": [row.as_dict() for row in goals],
        "groups": {
            group: [row.id for row in goals if row.group == group]
            for group in ("active", "paused", "complete", "set_aside")
        },
        "proposals": [row.as_dict() for row in proposals],
        "actions": ["draft", "propose", "confirm", "decline"],
    }


__all__ = ["plans"]
